i489-openshift-admin-guide (#518)
codenvy/che-docs: src/main/pages/setup-openshift/openshift-admin-guide.adoc

[id="examples"]
== Examples

All examples use the `kubectl` command. OpenShift administrators should use the `oc` command.

> [id="ram-prerequisites"]

Reviewer: @petra-sargent, did you check the whole body of docs for occurrences of the anchors you modified or removed?

    che-logs projects <volume1> <volume2>
----

Volumes can be anything that a user defines as volumes for workspace machines (volume name is equal to directory name in `${PV}/${ws-id}`).

When a workspace is deleted, a corresponding subdirectory (`${ws-id}`) is deleted in the PV directory.

[id="enabling-a-common-strategy"]
=== Enabling a common strategy

If you have already deployed Che with unique strategy:

* Set `CHE_INFRA_KUBERNETES_PVC_STRATEGY` to `common` in dc/che.

If applying the `che-server-template.yaml`:

* Pass `-p CHE_INFRA_KUBERNETES_PVC_STRATEGY=common` to the `oc new-app` command.

[NOTE]
====

> . For pre 1.6 Kubernetes, you need to set the `CHE_INFRA_KUBERNETES_PVC_PRECREATE__SUBPATHS` variable to `true`.

Reviewer: Don't know why you removed the pass-through macros, but they need to be there. Please, be sure to always check locally-built docs to spot such things (incorrect rendering without the macro).

```suggestion
. For pre 1.6 Kubernetes, you need to set the `pass:[CHE_INFRA_KUBERNETES_PVC_PRECREATE__SUBPATHS]` variable to `true`.
```

* If your infrastructure does not support workspace recovery, all running workspaces are forcibly stopped.

* If your infrastructure does support workspace recovery, any workspaces that are currently starting or stopping is allowed to finish that process. Running workspaces do not stop.

For those that did not stop, automatic fallback to the shutdown with full workspaces stopping will be performed.

If you want a full shutdown with workspaces stopped, you can request this by using the `shutdown=true` parameter. When preparation process is finished, the `READY_TO_SHUTDOWN` state is set which allows to stop current Che master instance.

[id="che-workspace-termination-grace-period"]
== Che workspace termination grace period

[IMPORTANT]
====
If the `terminationGracePeriodSeconds` variable is explicitly set in the OpenShift recipe, this environment variable does not override the recipe.
====

The default grace termination period of Kubernetes/OpenShift workspace’s pods is '0'.  This setting terminates pods almost instantly and significantly decreases the time required for stopping a workspace.

To increase the grace termination period:

> * Update `CHE_INFRA_KUBERNETES_POD_TERMINATION__GRACE__PERIOD__SEC`

Reviewer: ```suggestion
* Update `pass:[CHE_INFRA_KUBERNETES_POD_TERMINATION__GRACE__PERIOD__SEC]`.
```

[id="updating-che-without-stopping-active-workspaces"]
== Updating Che without stopping active workspaces

The differences between a Recreate update and a Rolling update:

Recreate update

* Che downtime

Rolling update

* No Che downtime
* New deployment starts in parallel and traffic is hot-switched

[id="performing-a-recreate-update"]
=== Performing a recreate update

Prerequisites

* Ensure that the new master version is fully API compatible with the old workspace agent version.

To perform a recreate update:

* Set the deployment update strategy to Recreate
* Make POST request to the /api/system/stop api to start WS master suspend (means that all new attempts to start workspaces will be refused, and all current starts/stops will be finished). Note that this method requires system admin credentials.

> * Make periodical GET requests to /api/system/state api, until it returns READY_TO_SHUTDOWN state. Also, you can check for "System is ready to shutdown" in the server logs.

Reviewer: ```suggestion
* Make periodical `GET` requests to the `/api/system/state` API, until it returns the `READY_TO_SHUTDOWN` state. Also, you can check for `System is ready to shutdown` in the server logs.
```

To delete Che server PVC, ServiceAccount and RoleBinding:
----
oc delete sa -l=app=che
oc delete rolebinding -l=app=che
----

To delete Keycloak and PostgreSQL PVCs
----
oc delete pvc -l=app=keycloak
oc delete pvc -l=app=postgres
----

[id="create-workspace-objects-in-personal-namespaces"]
== Creating workspace objects in personal namespaces

You can register the OpenShift server as an identity provider when Che is installed in multi-user mode. This allows you to create workspace objects in the OpenShift namespace of the user that is logged in Che through Keycloak.

To create a workspace object in the namespace of the user that is logged into Che:

* link:#openshift-identity-provider-registration[Register], inside Keycloak, an OpenShift identity provider that points to the OpenShift console of the cluster.
* link:#che-configuration[Configure] Che to use the Keycloak identity provider to retrieve the OpenShift tokens of the Che users.

Every workspace action such as start or stop creates an OpenShift resource in the OpenShift user account.  A notification message displays which allows you to link the Keycloak account to your OpenShift user account.

> But for non-interactive workspace actions, such as workspace stop on idling or Che server shutdown, the dedicated OpenShift account configured for the Kubernetes infrastructure is used. See link:admin-guide#who-creates-workspace-objects[AdminGuide] for more information.

Reviewer: ```suggestion
But for non-interactive workspace actions, such as workspace stop on idling or Che server shutdown, the dedicated OpenShift account configured for the Kubernetes infrastructure is used. See link:#setting-up-the-project-workspace[Setting up the project workspace] for more information.
```

To easily install Che on OpenShift with this feature enabled, see link:openshift-multi-user.html#creating-workspace-resources-in-personal-openshift-accounts-on-minishift[this section for Minishift] and link:openshift-multi-user.html#creating-workspace-resources-in-personal-openshift-accounts[this one for OCP]

[id="openshift-identity-provider-registration"]
==== OpenShift identity provider registration

[NOTE]
====
Cluster-wide administrator rights are required to add an OAuth client.
====

To add the OpenShift identity provider:

> . Use the following settings in the link:user-management#auth-and-user-management[Keycloak administration console]:

Reviewer: ```suggestion
. Use the following settings in the link:user-management.html#auth-and-user-management[Keycloak administration console]:
```

')
----

See https://www.keycloak.org/docs/3.3/server_admin/topics/identity-broker/social/openshift.html[Keycloak documentation] for more information on the Keycloak OpenShift identity provider.

[id="che-configuration"]
==== Che configuration

To configure Che deployment:

* Set the `CHE_INFRA_OPENSHIFT_PROJECT` environment variable to `NULL` to ensure a new distinct OpenShift namespace is created for every started workspace.

> * Set the `CHE_INFRA_OPENSHIFT_OAUTH__IDENTITY__PROVIDER` environment variable to the alias of the OpenShift identity provider specified in step 1 of its link:#openshift-identity-provider-registration[registration in Keycloak]. The default value is `openshift-v3`.

Reviewer: ```suggestion
* Set the `pass:[CHE_INFRA_OPENSHIFT_OAUTH__IDENTITY__PROVIDER]` environment variable to the alias of the OpenShift identity provider specified in step 1 of its link:#openshift-identity-provider-registration[registration in Keycloak]. The default value is `openshift-v3`.
```

[id="providing-the-openshift-certificate-to-keycloak"]
==== Providing the OpenShift certificate to Keycloak

If the certificate used by the OpenShift console is self-signed or is not trusted, then by default the Keycloak will not be able to contact the OpenShift console to retrieve linked tokens.

> Keycloak can not contact the OpenShift console to retrieve linked tokens when the certificate used by the OpenShift console is self-signed or is not trusted.

Reviewer: ```suggestion
Keycloak cannot contact the OpenShift console to retrieve linked tokens when the certificate used by the OpenShift console is self-signed or is not trusted.
```

> When the certificate is self-signed or is not trusted, use the enivronment variable `OPENSHIFT_IDENTITY_PROVIDER_CERTIFICATE` to pass the OpenShift console certificate to the Keycloak deployment. This will enable the Keycloak server to add the certificate to the list of trusted certificates.

Reviewer: ```suggestion
When the certificate is self-signed or is not trusted, use the `OPENSHIFT_IDENTITY_PROVIDER_CERTIFICATE` environment variable to pass the OpenShift console certificate to the Keycloak deployment. This will enable the Keycloak server to add the certificate to the list of trusted certificates.
```
